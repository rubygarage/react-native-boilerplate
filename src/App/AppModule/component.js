import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { SafeAreaProvider } from 'react-native-safe-area-context';

import AppNavigator from '../navigation/AppNavigator';
import ModalRoot from '../../views/ModalRoot';

const AppModuleComponent = ({ theme }) => (
  <ThemeProvider theme={theme}>
    <SafeAreaProvider>
      <AppNavigator />
      <ModalRoot />
    </SafeAreaProvider>
  </ThemeProvider>
);

AppModuleComponent.propTypes = {
  theme: PropTypes.shape().isRequired,
};

export default AppModuleComponent;
