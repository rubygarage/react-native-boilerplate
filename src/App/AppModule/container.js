import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { AppState } from 'react-native';
import { connect } from 'react-redux';

import theme from 'constants/theme';
import {
  startListenConnectionState as startListenConnectionStateAction,
  stopListenConnectionState as stopListenConnectionStateAction,
} from 'state/concepts/application/actions';
import AppModuleComponent from './component';

class AppModule extends Component {
  componentDidMount() {
    const { startListenConnectionState } = this.props;
    startListenConnectionState();
    AppState.addEventListener('change', this.handleAppStateChange);
  }

  componentWillUnmount() {
    AppState.removeEventListener('change', this.handleAppStateChange);
  }

  handleAppStateChange = (nextAppState) => {
    const { startListenConnectionState, stopListenConnectionState } = this.props;
    if (nextAppState === 'active') {
      startListenConnectionState();
    } else {
      stopListenConnectionState();
    }
  };

  render() {
    return (<AppModuleComponent theme={theme} />);
  }
}

AppModule.propTypes = {
  startListenConnectionState: PropTypes.func.isRequired,
  stopListenConnectionState: PropTypes.func.isRequired,
};

const mapDispatchToProps = {
  startListenConnectionState: startListenConnectionStateAction,
  stopListenConnectionState: stopListenConnectionStateAction,
};

export { AppModule as AppModuleContainer };
export default connect(null, mapDispatchToProps)(AppModule);
