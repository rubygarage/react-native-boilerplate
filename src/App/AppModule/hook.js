import { useEffect } from 'react';
import { AppState } from 'react-native';

import { useDispatch } from 'react-redux';

import {
  startListenConnectionState,
  stopListenConnectionState,
} from 'state/concepts/application/actions';

const useContainer = () => {
  const dispatch = useDispatch();

  /**
   * Handle App state.
   */
  const handleAppStateChange = (nextAppState) => {
    if (nextAppState === 'active') {
      dispatch(startListenConnectionState());
    } else {
      dispatch(stopListenConnectionState());
    }
  };

  // Did mount
  useEffect(() => {
    dispatch(startListenConnectionState());
    AppState.addEventListener('change', handleAppStateChange);

    return () => {
      AppState.removeEventListener('change', handleAppStateChange);
    };
  }, []);

  return {
    handleAppStateChange,
  };
};

export default useContainer;
