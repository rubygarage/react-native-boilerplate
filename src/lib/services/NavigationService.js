import { NavigationActions, StackActions } from 'react-navigation';

let navigator;

export const setTopLevelNavigator = (navigatorRef) => {
  navigator = navigatorRef;
};

export const navigate = (routeName, params) => {
  navigator.dispatch(
    NavigationActions.navigate({
      routeName,
      params,
    }),
  );
};

export const back = () => {
  navigator.dispatch(
    NavigationActions.back(),
  );
};

export const replace = (routeName, params) => {
  navigator.dispatch(
    StackActions.replace({ routeName, params }),
  );
};

export const reset = (index, actions, key) => {
  const resetAction = StackActions.reset({ index, actions, key });
  navigator.dispatch(resetAction);
};
