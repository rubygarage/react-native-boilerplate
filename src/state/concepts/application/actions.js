import * as types from './types';

export const setOnboardingCompletion = (isCompleted) => ({
  type: types.SET_ONBOARDING_COMPLETION,
  isCompleted,
});

export const startListenConnectionState = () => ({
  type: types.START_LISTEN_CONNECTION_STATE,
});

export const stopListenConnectionState = () => ({
  type: types.STOP_LISTEN_CONNECTION_STATE,
});

export const setConnectionState = (isConnected) => ({
  type: types.SET_CONNECTION_STATE,
  isConnected,
});
