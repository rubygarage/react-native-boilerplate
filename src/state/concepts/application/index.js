export { default } from './reducer';
export { default as applicationOperations } from './operations';
