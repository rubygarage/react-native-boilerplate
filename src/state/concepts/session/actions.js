import * as types from './types';

export const setTokens = (tokens) => ({
  type: types.SET_TOKENS,
  tokens,
});

export const removeTokens = () => ({
  type: types.SET_TOKENS,
  tokens: null,
});

export const setUserId = (id) => ({
  type: types.SET_USER_ID,
  id,
});
