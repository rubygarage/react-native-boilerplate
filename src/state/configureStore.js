import { createStore, applyMiddleware } from 'redux';
import { createLogicMiddleware } from 'redux-logic';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { persistStore, persistReducer } from 'redux-persist';
import { Settings } from 'luxon';
import AsyncStorage from '@react-native-community/async-storage';

import buildHttpClient from 'lib/httpClient';
import { rootHttpClient } from 'lib/httpClient/buildHttpClient';
import mountInterceptors from 'lib/httpClient/mountInterceptors';
import conceptsOperations from './concepts/operationsRoot';
import flashOperations from './flash/operations';
import rootReducer from './store';

const operations = [...conceptsOperations, ...flashOperations];

const configureStore = () => {
  const persistConfig = {
    key: 'root',
    storage: AsyncStorage,
    whitelist: ['data', 'application'],
    timeout: null,
  };
  const persistedReducer = persistReducer(persistConfig, rootReducer);

  const operationsDependencies = {
    httpClient: buildHttpClient(),
    rootHttpClient,
  };

  const logicMiddleware = createLogicMiddleware(operations, operationsDependencies);
  const middlewares = applyMiddleware(logicMiddleware);
  const enhancer = composeWithDevTools(middlewares);

  const store = createStore(persistedReducer, enhancer);
  const persistor = persistStore(store);
  store.logicMiddleware = logicMiddleware;
  store.httpClient = operationsDependencies.httpClient;

  mountInterceptors(store);

  if (module.hot) {
    module.hot.accept(() => {
      // eslint-disable-next-line global-require
      store.replaceReducer(require('./store').default);
    });
  }

  Settings.defaultLocale = store.getState().intl.locale;

  return { store, persistor };
};

export default configureStore;
