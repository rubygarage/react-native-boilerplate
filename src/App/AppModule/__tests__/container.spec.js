import React from 'react';
import { AppState } from 'react-native';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import diveTo from 'utils/testHelpers/diveToEnzyme';
import { setTopLevelNavigator } from 'lib/services/NavigationService';
import { startListenConnectionState, stopListenConnectionState } from 'state/concepts/application/actions';
import AppModule, { AppModuleContainer } from '../container';

jest.spyOn(AppState, 'addEventListener');
jest.spyOn(AppState, 'removeEventListener');

jest.mock('lib/services/NavigationService', () => ({
  setTopLevelNavigator: jest.fn(),
}));

jest.mock('state/concepts/application/actions', () => ({
  startListenConnectionState: jest.fn(),
  stopListenConnectionState: jest.fn(),
}));

const store = configureStore()({});
store.dispatch = jest.fn();
const defaultProps = { store };
const testRef = {};

describe('App module', () => {
  const wrapper = shallow(<AppModule {...defaultProps} />);
  const container = diveTo(wrapper, AppModuleContainer);
  const instance = container.instance();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('componentDidMount listen deep links', () => {
    instance.componentDidMount();
    expect(startListenConnectionState).toHaveBeenCalled();
    expect(AppState.addEventListener).toHaveBeenCalledWith('change', instance.handleAppStateChange);
  });

  it('componentWillUnmount removes subscription', () => {
    instance.componentWillUnmount();
    expect(AppState.removeEventListener).toHaveBeenCalledWith('change', instance.handleAppStateChange);
  });

  it('calls startListenConnectionState', async () => {
    const appState = 'active';
    await instance.handleAppStateChange(appState);
    expect(startListenConnectionState).toHaveBeenCalled();
  });

  it('calls stopListenConnectionState', () => {
    const appState = 'background';
    instance.handleAppStateChange(appState);
    expect(stopListenConnectionState).toHaveBeenCalled();
  });

  it('handleNavRef calls setTopLevelNavigator', () => {
    instance.handleNavRef(testRef);
    expect(setTopLevelNavigator).toHaveBeenCalledWith(testRef);
  });
});
