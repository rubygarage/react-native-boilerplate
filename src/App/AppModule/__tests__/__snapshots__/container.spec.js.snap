// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App module renders correctly 1`] = `
<AppModuleComponent
  theme={
    Object {
      "colors": Object {
        "black": "#000000",
        "white": "#ffffff",
      },
      "fonts": Object {},
      "sizes": Object {},
    }
  }
/>
`;
